File: fasta/fasta_size_report.pl
#!/usr/bin/env perl

=head1 NAME

fasta_size_report.pl - print a file containing fasta ID, sequence size, the defline, 
and source file (NOTE: Only verified on a single file input)

=head1 SYNOPSIS

USAGE: fasta_size_report.pl 
            --file=/path/to/some_file.fsa --file=/path/to/some/other/file.fsa
                       OR
            --list=/path/to/some_file.list
                       OR
            --dir=/path/to/somedir
          [ --summary=1 ]

=head1 OPTIONS

B<--file,-f>
    a single or multiple fasta files of interest

B<--list,-l>
    an input list containing a set of fasta files

B<--dir,-d>
    a directory containing .fsa, .fasta, .fa files 

B<--summary,-s>
    optional.  If passed, also exports summary lines (each begins with a # symbol)

B<--log,-l> 
    Log file

B<--help,-h>
    This help message

=head1  DESCRIPTION

This script takes an input fasta file, a file containing a list of fasta files, or a 
directory of fasta files and prints a synopsis file containing the ID, the size of the 
sequence, the original defline, and the originating file.

=head1  INPUT

There are several potential inputs:

- A single fasta file

- A list of fasta files:
    
    fasta.list
      /path/to/A.fasta
      /path/to/B.fasta

- A directory of fasta files with the extension .fsa, .fa, or .fasta

=head1  OUTPUT

The output is a tabular list of:

ID    SIZE    DEFLINE    FILE
1     100     >1 blah    /path/to/A.fasta

Optionally, if the -s option is set to 1, also exports summary lines, each beginning
with a comment.

=head1  CONTACT

    Anu Ganapathy
    alenas@gmail.com

=cut

use warnings;
use strict;
use Getopt::Long qw(:config no_ignore_case no_auto_abbrev pass_through);
use Pod::Usage;




my %options = ();
my $results = GetOptions (\%options, 
                          'file|f=s@',
                          'list|l=s',
                          'dir|d=s',
                          'summary|s',
                          'log|l=s',
                          'help|h') || pod2usage();

## display documentation
if( $options{'help'} ){
    pod2usage( {-exitval => 0, -verbose => 2, -output => \*STDERR} );
}

## open the log if requested
my $logfh;
if (defined $options{log}) {
    open($logfh, ">$options{log}") || die "can't create log file: $!";
}

## make sure everything passed was peachy and retrieve files of interest
my @files = &check_parameters(\%options);

my $summary_record_count = 0;
my $summary_size = 0;

## print mapping for each file
foreach my $file (@files){
    &print_size($file);
}

if ( $options{summary} ) {
    print "\nSummary:\n" .
          "\tRecord count   : $summary_record_count\n" .
          "\tSum of residues: $summary_size\n";
}

exit(0);


sub _log {
    my $msg = shift;

    print $logfh "$msg\n" if $logfh;
}


sub check_parameters {
    my $options = shift;
    
    ## make sure required arguments were passed
    my @input_types = qw( file dir list );
    my @inputs = ();

    ## file
    if  ( defined $options->{file} ) {
        @inputs = @{$options->{file}};
    }

    ## list
    if ( defined $options->{list} ) {
	open FILE, $options->{list} || die "Could not open input list $options->{list}: $!";
	while (my $file = <FILE>){
	    chomp;
	    push @inputs, $file;
	}
	close FILE;
    }

    ## dir (accept all 3 ways specified in documentation)
    if( defined $options->{dir} ) {
        die "Invalid path give for directory input: $options->{dir} " if (! -d $options->{dir} );
        my $dir = $options->{dir};
        my @files = <$dir/*.fsa>;
        @inputs = (@inputs, @files);
        @files = <$dir/*.fa>;
        @inputs = (@inputs, @files);
        @files = <$dir/*.fasta>;
        @inputs = (@inputs, @files);
    }

    ## check to make sure we have files 
    die "Script requires at least one fasta file input via the -f or -l options" if (scalar @inputs == 0);

    ## check to make sure they all exist
    foreach my $file (@inputs){
        die "Error with file $file.  Check if it exists and is readable" if (!-e $file || !-r $file);
    }

    return @inputs;
}

sub print_size {
    my ($file) = @_;

    open (my $FIN, $file) || die "Unable to read fasta file ($file): $!";

    ## initialize fields of interest
    my $len     = 0; 
    my $defline = "";
    my $id      = "";
    
    ## go through file
    while (my $line = <$FIN>) {
	chomp($line);
	
	## header line
	if ($line =~ /^>/) {

	    ## if we processed a line already
	    if ($len > 0) {
		print "$id\t$len\t$defline\t$file\n";
        $summary_record_count++;
        $summary_size += $len;
        $len = 0;
	    }

	    $defline = $line;

	    $line =~ s/^>//;
	    ($line) = split(/\s+/, $line);
	    $id = $line;
	}
	## sequence line
	else {
	    $line =~ s/\s+//g;  ## remove spaces
	    $len += length($line);
	}
    }
    print "$id\t$len\t$defline\t$file\n";
    $summary_record_count++;
    $summary_size += $len;


    close($FIN);

}

File: fasta/fasta_size_report.py
#!/usr/bin/env python3

'''

Description:

    Accepts an input fasta file(s) and prints a synopsis file containing the ID, the size of the sequence, the original defline, and the originating file.


Usage Options:

    --file,-f
        Single or multiple fasta files of interest
        -f /path/to/file.fa /path/to/file.fa

    --list,-l
        Input list containing a set of fasta files
        -l /path/to/list.txt

    --dir,-d
        A directory containing .fsa, .fasta, .fa files
        -d /path/to/fa_files/

    --summary,-s    (Optional)
        If passed, also exports summary metrics

    --help,-h
        This help message

Output:
    Tab delimited list of fasta file info like following example. Optionally, if '-s' is set it also exports summary metrics.

    ID    SIZE    DEFLINE    FILE
    1     100     >1 blah    /path/to/A.fasta

Original Perl Author:

    Anu Ganapathy
    alenas@gmail.com

Python Translation:

    Chris
    RPINerd@gmail.com

'''

import argparse
import os


def main(args):

    # Convert file argument into parsable list
    file_list = handle_file_input(args)

    # Process input fasta file(s)
    records = []
    for fa in file_list:

        # Check for valid file, but skip gracefully if missing
        file = f"{os.getcwd()}/{str(fa)}"
        if not os.path.exists(file):
            print(f"The fasta file {file} was not found! Skipping..")
            continue

        records.extend(parse_file(file))

    # Print metrics
    print("\t".join(["ID", "SIZE", "DEFLINE", "FILE"]))
    total_records = len(records)
    total_residues = 0
    for record in records:
        total_residues += int(record[1])
        print("\t".join(record))

    if args.summary:
        print(f"Summary:\n\tRecords:\t{total_records}\n\tResidues:\t{total_residues}")


# Given a single file, parse and calculate metrics
def parse_file(fasta):

    record = []
    with open(fasta, "r") as fa_file:
        short_file = fasta.split("/")[-1]
        length = 0
        defline = ""
        id = ""

        for line in fa_file:
            line = line.strip()

            if line.startswith(">"):

                if length > 0:
                    record.append([id, str(length), defline, short_file])
                    length = 0

                defline = str(line)

                id = line.lstrip(">").split(" ")[0]
            else:
                length += len(line)

        record.append([id, str(length), defline, short_file])

    return record


# Based on input type, builds and returns list of file names
def handle_file_input(args):

    file_list = []

    if args.dir:

        assert os.path.exists(args.dir), f"Directory not found at location {args.dir}"

        for file in os.listdir(args.dir):
            if file.endswith((".fa", ".fasta")):
                file_list.append(file)

    elif args.list:

        assert os.path.exists(args.list), f"List file {args.list} was not found!"

        with open(args.list, "r") as fa_list:
            for line in fa_list:
                file_list.append(line.strip())

    else:
        file_list.extend(args.file.split(" "))

    assert len(file_list) > 0, "At least one input file is required!"

    return file_list


if __name__ == '__main__':

    parser = argparse.ArgumentParser(description='Prints a synopsis file containing the ID, the size of the sequence, the original defline, and the originating file')

    # Input file options
    input_files = parser.add_mutually_exclusive_group(required=True)
    input_files.add_argument('-f', '--file', type=str, help='Path to an input file to be read' )
    input_files.add_argument('-l', '--list', type=str, help='Path to a file containing a list of fastas' )
    input_files.add_argument('-d', '--dir', type=str, help='Path to a directory containing fasta files' )

    # Optional flags
    parser.add_argument('-s', '--summary', required=False, help='Generate summary lines in report', action="store_true")

    args = parser.parse_args()
    main(args)
